arangod/RocksDBEngine/RocksDBPersistedLog.cpp

```diff
@@ -0,0 +1,290 @@
+
+  if (options.waitForSync) {
+    // At this point we have to make sure that every previous log entry is
+    // synced as well. Otherwise we might get holes in the log.
+    if (auto s = _persistor->_db->SyncWAL(); !s.ok()) {
+      return rocksutils::convertStatus(s);
+    }
+  }
+
+  return {};
+}
+
+struct RocksDBLogIterator : replication2::LogIterator {
+  ~RocksDBLogIterator() override = default;
+
+  RocksDBLogIterator(RocksDBPersistedLog* log, rocksdb::DB* db,
+                     rocksdb::ColumnFamilyHandle* cf, LogIndex start)
+      : _bounds(log->getBounds()), _upperBound(_bounds.end()) {
+    rocksdb::ReadOptions opts;
+    opts.prefix_same_as_start = true;
+    opts.iterate_upper_bound = &_upperBound;
+    _iter.reset(db->NewIterator(opts, cf));
+
+    auto first = RocksDBKey{};
+    first.constructLogEntry(log->objectId(), start);
+    _iter->Seek(first.string());
+  }
+
+  auto next() -> std::optional<LogEntry> override {
+    if (!_first) {
+      _iter->Next();
+    }
+    _first = false;
+
+    if (!_iter->Valid()) {
+      auto s = _iter->status();
+      if (!s.ok()) {
+        auto res = rocksutils::convertStatus(s);
+        THROW_ARANGO_EXCEPTION_MESSAGE(res.errorNumber(), res.errorMessage());
+      }
+
+      return std::nullopt;
+    }
+
+    return std::optional<LogEntry>{std::in_place, RocksDBValue::logTerm(_iter->value()),
+                                   RocksDBKey::logIndex(_iter->key()),
+                                   RocksDBValue::logPayload(_iter->value())};
+  }
+
+  RocksDBKeyBounds const _bounds;
+  rocksdb::Slice const _upperBound;
+  std::unique_ptr<rocksdb::Iterator> _iter;
+  bool _first = true;
+};
+
+auto RocksDBPersistedLog::read(LogIndex start) -> std::unique_ptr<LogIterator> {
+  return std::make_unique<RocksDBLogIterator>(this, _persistor->_db, _persistor->_cf, start);
+}
+
+auto RocksDBPersistedLog::drop() -> Result {
+  TRI_ASSERT(false);
+  THROW_ARANGO_EXCEPTION(TRI_ERROR_NOT_IMPLEMENTED);
+}
+
+auto RocksDBPersistedLog::removeFront(replication2::LogIndex stop) -> Result {
+  auto last = RocksDBKey();
+  last.constructLogEntry(_objectId, LogIndex{stop.value});
+
+  rocksdb::WriteOptions opts;
+  auto s = _persistor->_db->DeleteRange(opts, _persistor->_cf,
+                                        getBounds().start(), last.string());
+  return rocksutils::convertStatus(s);
+}
+
+auto RocksDBPersistedLog::removeBack(replication2::LogIndex start) -> Result {
+  auto first = RocksDBKey();
+  first.constructLogEntry(_objectId, LogIndex{start.value});
+
+  rocksdb::WriteOptions opts;
+  auto s = _persistor->_db->DeleteRange(opts, _persistor->_cf, first.string(),
+                                        getBounds().end());
+  return rocksutils::convertStatus(s);
+}
+
+auto RocksDBPersistedLog::prepareWriteBatch(replication2::LogIterator& iter,
+                                            rocksdb::WriteBatch& wb) -> Result {
+  while (auto entry = iter.next()) {
+    auto key = RocksDBKey{};
```

Comment:
The key should be declared outside the loop, so it can be recycled for each log entry.